### PR TITLE
Advance checkpoints only after persisting ops

### PR DESCRIPTION
**Pull request by marregui**
Do-over

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by marregui** [comment]
retest this please

cannot reproduce locally

File: es/es-server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 1]
```diff
@@ -1368,19 +1373,38 @@ public long getIfPrimaryTerm() {
...
+            this(true, docIdAndVersion.version, docIdAndVersion, searcher);
+        }
+
+        public boolean exists() {
+            return exists;
+        }
+
+        public long version() {
+            return this.version;
+        }
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/engine/Engine.java]: Can we remove these extra properties again? See https://github.com/crate/crate/commit/836b1bb176b3c1ce4116e2d812597b37da8a6740 for why they were removed.

File: es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -186,17 +189,25 @@ public InternalEngine(EngineConfig engineConfig) {
...
-
             mergeScheduler = scheduler = new EngineMergeScheduler(engineConfig.getShardId(), engineConfig.getIndexSettings());
             throttle = new IndexThrottle();
             try {
-                translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier());
+                trimUnsafeCommits(engineConfig);
+                translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier(),
+                    seqNo -> {
+                        final LocalCheckpointTracker tracker = getLocalCheckpointTracker();
+                        assert tracker != null || getTranslog().isOpen() == false;
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Could we add some assertion message here?

File: es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -243,29 +252,25 @@ public InternalEngine(EngineConfig engineConfig) {
...
+
+    private LocalCheckpointTracker createLocalCheckpointTracker(
+        BiFunction<Long, Long, LocalCheckpointTracker> localCheckpointTrackerSupplier) throws IOException {
+        final long maxSeqNo;
+        final long localCheckpoint;
+        final SequenceNumbers.CommitInfo seqNoStats =
+            SequenceNumbers.loadSeqNoInfoFromLuceneCommit(store.readLastCommittedSegmentsInfo().userData.entrySet());
+        maxSeqNo = seqNoStats.maxSeqNo;
+        localCheckpoint = seqNoStats.localCheckpoint;
+        logger.trace("recovered maximum sequence number [{}] and local checkpoint [{}]", maxSeqNo, localCheckpoint);
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: maybe wrap in `logger.isTraceEnabled`

File: es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -888,8 +900,11 @@ public IndexResult index(Index index) throws IOException {
                     versionMap.maybePutIndexUnderLock(index.uid().bytes(),
                         new IndexVersionValue(translogLocation, plan.versionForIndexing, plan.seqNoForIndexing, index.primaryTerm()));
                 }
-                if (indexResult.getSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO) {
-                    localCheckpointTracker.markSeqNoAsCompleted(indexResult.getSeqNo());
+                localCheckpointTracker.markSeqNoAsProcessed(indexResult.getSeqNo());
+                if (indexResult.getTranslogLocation() == null) {
+                    // the op is coming from the translog (and is hence persisted already) or it does not have a sequence number
+                    assert index.origin().isFromTranslog() || indexResult.getSeqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO;
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Same here, an assertion message would be nice

File: es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 5]
```diff
@@ -1261,8 +1276,11 @@ public DeleteResult delete(Delete delete) throws IOException {
                 final Translog.Location location = translog.add(new Translog.Delete(delete, deleteResult));
                 deleteResult.setTranslogLocation(location);
             }
-            if (deleteResult.getSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO) {
-                localCheckpointTracker.markSeqNoAsCompleted(deleteResult.getSeqNo());
+            localCheckpointTracker.markSeqNoAsProcessed(deleteResult.getSeqNo());
+            if (deleteResult.getTranslogLocation() == null) {
+                // the op is coming from the translog (and is hence persisted already) or does not have a sequence number (version conflict)
+                assert delete.origin().isFromTranslog() || deleteResult.getSeqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO;
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: and here.

**Comment by marregui** [comment]
retest this please

I cannot reproduce locally....

**Comment by marregui** [comment]
retest this please

seems flaky, I am investigating locally

**Comment by marregui** [comment]
retest this please

it is flaky, and I cannot reproduce it locally. Taking the one from ES requires some changes, some code to import too

File: es/es-server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 6]
```diff
@@ -4442,6 +4458,8 @@ public void testAcquireIndexCommit() throws Exception {
         }
     }
 
+    @Seed("EA3633FB2FE61702")
+    @Repeat(iterations = 1)
```
Comment (mfussenegger) [inline es/es-server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I guess this should be reverted now

File: es/es-server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 7]
```diff
@@ -5054,6 +5076,8 @@ public void testTrackMaxSeqNoOfUpdatesOrDeletesOnPrimary() throws Exception {
         }
     }
 
+    @Seed("99FCB15FF5D32AD0")
+    @Repeat(iterations = 1)
```
Comment (mfussenegger) [inline es/es-server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: Same here

**Pull request merged**